### ISSUE TITLE
Overload on constants: type checking additional parameters

### ISSUE THREAD

**Issue by NoelAbrahams** [Microsoft/TypeScript#191]
Hi, consider this bit of code:

```typescript
class Bar {
    public bar: string;
}
interface IFoo {
    doFoo(type: any, bar: any): void;
    doFoo(type: "Quux", bar: Bar): string;
}
```

We use it like this:
```typescript
var foo: IFoo;
var result:string = foo.doFoo("Quux", new Bar()); // fine
```

However, while still specifying "Quux", if we pass any type other than new Bar() for the second argument, the compiler falls back to the non-overloaded method:
```typescript
// Error TS2011: Cannot convert 'void' to 'string'.
var result2: string = foo.doFoo("Quux", "bar");
```

I think this is a bit unfortunate. We are missing an opportunity to verify that the additional arguments also match the overloaded signature.

Ideally the following
```typescript
foo.doFoo("Quux", "bar");
```
should generate error TS2082: Supplied parameters do not match any signature of call target

See further discussion on the old [codeplex site](https://typescript.codeplex.com/workitem/993).

**Comment by mhegazy** [comment]
With https://github.com/Microsoft/TypeScript/pull/6278 in place, you can omit the general signature, and the expected checks should take place. also as noted in https://github.com/Microsoft/TypeScript/issues/6746#issuecomment-186380536, doing so will restrict how the function can be called.

**Comment by NoelAbrahams** [comment]
This is great news. We've had to hack around this for a long time. I hope the fix makes it into a release soon. Thanks.
